packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -1602,3 +1602,160 @@ export class Addr extends Op {
+   * @param line line number in TEAL file
+   * @param interpreter Interpreter Object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super();
+    this.interpreter = interpreter;
+
+    assertLen(args.length, 0, line);
+  };
+
+  execute (stack: TEALStack): void {
+    this.assertMinStackLen(stack, 1);
+    let last = this.assertBigInt(stack.pop());
+    let bal = 0;
+
+    if (last === BigInt("0")) {
+      const val = this.interpreter.accounts.get(convertToString(this.interpreter.tx.snd));
+      if (val === undefined) {
+        throw new TealError(ERRORS.TEAL.ACCOUNT_DOES_NOT_EXIST);
+      }
+      bal = val.amount;
+    } else {
+      this.checkIndexBound(Number(--last), this.interpreter.tx.apat);
+
+      const buf = this.interpreter.tx.apat[Number(last)];
+      const res = this.interpreter.accounts.get(convertToString(buf));
+
+      if (res === undefined) {
+        throw new TealError(ERRORS.TEAL.ACCOUNT_DOES_NOT_EXIST);
+      }
+
+      bal = res.amount;
+    }
+
+    stack.push(BigInt(bal));
+  }
+}
+
+// get Asset Holding Info for given account
+// read from account specified by Txn.Accounts[A]
+// and asset B holding field X (imm arg) => {0 or 1 (top), value}
+// params: account index, asset id. Return: did_exist flag (1 if exist and 0 otherwise), value.
+export class AssetHoldingGet extends Op {
+  readonly interpreter: Interpreter;
+  readonly field: string;
+
+  /**
+   * @param args words list extracted from line: {Asset Holding field}
+   * @param line line number in TEAL file
+   * @param interpreter Interpreter Object
+   */
+  constructor (args: string[], line: number, interpreter: Interpreter) {
+    super();
+    this.interpreter = interpreter;
+    assertLen(args.length, 1, line);
+
+    this.field = args[0];
+  };
+
+  execute (stack: TEALStack): void {
+    this.assertMinStackLen(stack, 2);
+    const last = this.assertBigInt(stack.pop());
+    let prev = this.assertBigInt(stack.pop());
+
+    this.checkIndexBound(Number(--prev), this.interpreter.tx.apat);
+
+    const accBuffer = this.interpreter.tx.apat[Number(prev)];
+    const val = this.interpreter.accountsAssets.get(convertToString(accBuffer));
+    const res = val?.get(Number(last));
+
+    if (res === undefined) {
+      stack.push(BigInt("0"));
+    } else {
+      let value: StackElem;
+
+      switch (this.field) {
+        case "AssetBalance":
+          value = BigInt(res.amount);
+          break;
+        case "AssetFrozen":
+          value = toBytes(res["is-frozen"]);
+          break;
+        default:
+          throw new TealError(ERRORS.TEAL.INVALID_FIELD_TYPE);
+      }
+
+      stack.push(value);
+      stack.push(BigInt("1"));
+    }
+  }
+}
+
+// get Asset Params Info for given account
+// read from asset Txn.ForeignAssets[A] params field X (imm arg) => {0 or 1 (top), value}
+// params: txn.ForeignAssets offset. Return: did_exist flag (1 if exist and 0 otherwise), value.
```

Comment:
Please improve the comment. You can look at my suggestion of the operator above.

Reply:
ok, currently these comments are taken from TEAL documentation, updating